server: Create new api for table metadata

kyle-a-wong opened the pull request:
adds a new api v2 endpoint, `/api/v2/table_metadata/`.

This new API enables the ability to retrieve paginated table metadata for a given database.

Resolves: https://github.com/cockroachdb/cockroach/issues/128894
Epic: [CRDB-37558](https://cockroachlabs.atlassian.net/browse/CRDB-37558)
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128231)
<!-- Reviewable:end -->

kyle-a-wong committed:
server: create new api for table metadata

Adds a new API v2 endpoint,
`/api/v2/table_metadata/`, That accepts
the following query parameters:
 - dbId
 - name
 - sortBy
 - sortOrder
 - pageSize
 - pageNum
 - storeId

This new API enables the ability to retrieve
table metadata, using the supported query
to filter, sort, and paginate the response.

The query params dbId, name, and storeId are
used to filter down the total results of the
table_metadata query.

The query params sortBy and sortOrder affect the
sorting of the total results, before pagination.

The query params pageSize and pageNum limit the
size of the result set returned to the client.

The user making the request to this new endpoint
will only see tables for databases they have the
`CONNECT` database privilege for. If the user has
the `ADMIN` privilege, they will be able to see
all tables in all databases.

Resolves: #128894
Epic: CRDB-37558
Release note: None

angles-n-daemons commented on pkg/server/apiutil/apiutil.go:
```diff
@@ -30,3 +32,59 @@ func WriteJSONResponse(ctx context.Context, w http.ResponseWriter, code int, pay
+	return QueryStringVal[[]int]{Exists: false}
+}
+
+type QueryStringVal[T any] struct {
```
Do you feel strongly about adding this struct? I feel that nil or zero-value checks should be done to indicate existence, eg:

```go
if c.tableName == nil {
```
or
```go
if c.tableName == "" {
```

angles-n-daemons commented on pkg/server/api_v2_databases_metadata.go:
```diff
@@ -0,0 +1,385 @@
+		pageNum   apiutil.QueryStringVal[int]
+		pageSize  apiutil.QueryStringVal[int]
+	}{
+		dbId:      apiutil.GetIntQueryStringVal(queryValues, dbIdKey),
```
What are your thoughts about inlining the utility functions into a function for this struct?

```go
type tableMetadataQueryArgs  struct {
     dbId            int
     tableName string
     ...
}

func newTableMetadataQueryArgs(vals url.Values) (tableMetadataQueryArgs, error) {
    // if you want to check zero values
    dbIdStr := vals.Get(dbIdKey)
    dbId, err := strconv.Atoi(dbIdStr)
    if err != nil {
        return errors.Wrapf(err, "could not parse tableMetadata query arg %s", dbIdKey)
    }

     // if nil checks work better to verify existence
    var tableName *string
    if vals.Has(nameKey) {
        name := vals.Get(nameKey)
        tableName = &name
    }

    storeIds := make([]int, len(vals[storeIds])
    for _, storeIdStr := range vals[storeIds] {
         storeId, err := strconv.Atoi(storeIdStr)
         if err != nil {
             return errs.Wrapf(err, "could not parse tableMetadata storeId: %s", storeIdStr)
         }
         storeIds = append(storeIds)
    }
   

    return tableMetadataQueryArgs{
        tableName: tableName,
        ...
    }
} 
```

It's a bit wordy, but I think more clearly defines how the query will be parsed (absent an api specification)

angles-n-daemons replied:
I see, yeah it makes sense since they're passed individually as arguments to the next function down - I think my only question is around the utility functions `GetIntQueryStringVal`. Do you have any thoughts about trying thing PR without them?

kyle-a-wong replied:
Done

kyle-a-wong replied:
Actually, I added GetIntQueryStringVal back in. I think not having this function results in a lot of duplicated code to get the param value and convert it to an int.

angles-n-daemons reviewed (commented):
couple of comments, this looks pretty good overall

shailendra-patel committed:
roachprod: upgrade go sdk for azure

Upgraded version for azure from v57.1.0 to  v68.0.0

Changes as part of this PR:
- Upgraded azure sdk to version v68.0.0.

Epic: none
Informs: https://github.com/cockroachdb/cockroach/issues/129009

Release note: None

angles-n-daemons commented on pkg/server/api_v2_databases_metadata.go:
```diff
@@ -0,0 +1,385 @@
+// if the user does not provide a database id, the result set will only include databases in which they have the
+// previously mentioned authorizations.
+//
+// This endpoint supports the following query parameters:
```
Oh I remember this discussion, definitely makes sense - feel free to proceed in either direction

angles-n-daemons commented on pkg/server/api_v2_databases_metadata.go:
```diff
@@ -0,0 +1,385 @@
+	ctx := r.Context()
+	ctx = a.sqlServer.AnnotateCtx(ctx)
+	sqlUser := authserver.UserFromHTTPAuthInfoContext(ctx)
+	// TODO (kyle): build http method handling directly into route registration
```
got it, I see that makes sense to me

xinhaoz reviewed (commented):
Nice! Question on the virtual table usage but looking great overall!

> Reviewed 4 of 120 files at r3, 2 of 11 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-:-O6Nx4Ai6_MOi5FgYe14:b9qyx5s)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @kyle-a-wong)

___
*[`-- commits` line 4 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O5TobPk5C2_Dw0kkB3j):*
I think this message can be polished a bit.
___
*[`pkg/server/api_v2_databases_metadata.go` line 145 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O5TrRsnE7GCO18C3K0c:-O5TrRsnE7GCO18C3K0d:bh7k7im) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea5105adf00fb28d2d0beff99d2fbeb6297d88e0/pkg/server/api_v2_databases_metadata.go#L145)):*
> ```Go
> 		}
> 
> 		switch qs.sortOrder.Value {
> ```

nit: Why do we need a switch here?
```
sortOrder := "ASC"
if qs.sortOrder.Value == "desc" {
  sortOrder = "DESC"
}
```
___
*[`pkg/server/api_v2_databases_metadata.go` line 180 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O6MKlEg1j-hibmAwi0x:-O6MKlEg1j-hibmAwi0y:bwv9wht) ([raw file](https://github.com/cockroachdb/cockroach/blob/efbf2dbad82304b4b4625020d9c7044596c055ee/pkg/server/api_v2_databases_metadata.go#L180)):*
> ```Go
> 	var sortOrder string
> 	if sortBy != "" {
> 		switch sortOrderQS {
> ```

nit: Do we need a switch here?
```
var sortOrder string
if sortBy != "" {
    sortOrder = "ASC"
    if sortOrderQS == "desc" {
        sortOrder = "DESC"
    }
}
```
___
*[`pkg/server/api_v2_databases_metadata.go` line 252 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O6MO3rg2Q2UYVwGnmJt:-O6MO3rg2Q2UYVwGnmJu:b3t2it8) ([raw file](https://github.com/cockroachdb/cockroach/blob/efbf2dbad82304b4b4625020d9c7044596c055ee/pkg/server/api_v2_databases_metadata.go#L252)):*
> ```Go
> 			count(*) OVER() as total_row_count
> 		FROM system.table_metadata tbm
> 		JOIN crdb_internal.databases dbs ON dbs.id = tbm.db_id
> ```

Do we know if the usage of this virtual table scales well to large clusters / won't hit perf issues with a lot of descriptors?


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> Reviewed 2 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128231)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @xinhaoz)

___
*[`pkg/server/api_v2_databases_metadata.go` line 180 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O6MKlEg1j-hibmAwi0x:-O6WRtxy0BGbBJ7IGfsE:b-n9tkhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/efbf2dbad82304b4b4625020d9c7044596c055ee/pkg/server/api_v2_databases_metadata.go#L180)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

nit: Do we need a switch here?
```
var sortOrder string
if sortBy != "" {
    sortOrder = "ASC"
    if sortOrderQS == "desc" {
        sortOrder = "DESC"
    }
}
```
</blockquote></details>

changed
___
*[`pkg/server/api_v2_databases_metadata.go` line 252 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O6MO3rg2Q2UYVwGnmJt:-O6WRwbh-Jjlu0KXkFPU:b-hw79p1) ([raw file](https://github.com/cockroachdb/cockroach/blob/efbf2dbad82304b4b4625020d9c7044596c055ee/pkg/server/api_v2_databases_metadata.go#L252)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Do we know if the usage of this virtual table scales well to large clusters / won't hit perf issues with a lot of descriptors?
</blockquote></details>

I removed the usage of this virtual table as it isn't strictly necessary.  One thing that this extra join provided was protection against "losing" tables in the view in the case that they have been renamed, since the name will no longer match the cluster_database_privileges. But I'm not sure how frequent an action like this occurs, but I assume it isn't very frequent any probably isn't worth the extra join.

To answer the question, I'm not sure about the performance of this VT in large clusters.


<!-- Sent from Reviewable.io -->

xinhaoz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for addressing comments! Let's get these APIs merged to start connecting to the UI. I'm sure we'll need to iterate on them slightly anyways once that's in progress.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-:-O6WbVAV3nBwP3B8LZyL:b-mykh1r)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @kyle-a-wong)

___
*[`-- commits` line 2 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O6WbPhq3SJxYXu9usfZ):*
supernit: We also typically don't capitalize the title after the `pkgName:` portion.


<!-- Sent from Reviewable.io -->

angles-n-daemons reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> also lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-:-O6Wkfxr4Xz2WwPZ-SV1:b-s8fz3x)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @herkolategan, @kyle-a-wong, @renatolabs, and @xinhaoz)



<!-- Sent from Reviewable.io -->

kyle-a-wong commented:
TFTR!

bors r+
